### ISSUE TITLE
Handling per layer per parameter statistics in Neural Network optimizers

### ISSUE THREAD

**Issue by sumedhghaisas** [mlpack/mlpack#1017]

**Comment by sumedhghaisas** [comment]
The current mlpack ann framework accommodates a nice and elegant way to make the use of currently implemented optimizers in mlpack. Most optimizers update a single matrix based on their update rule. Therefore, the current ANN framework merges all the parameters of all the layers into 1 single parameter matrix. Although elegant, this procedure can be computationally expensive as many matrix transformations are required in transforming each layer's individual matrices to global parameter matrix and vice versa. The solution for this is to use a specialized update rule for ANN which encapsulates the original update rule and calls this original update rule on every parameter separately. Using such specialized update rules for some classes can be done with the use of traits and without involving a lot of code changes. 

Although the bigger problem lies in handling an update policy object for each parameter in the network. This update policy object might be holding parameter dependent stats for gradient update. Each parameter has to be updated with its associated update policy object. This requires some management framework.

Solution - (elegant and lazy): We can ask the layer to register their parameters to a global collector at initialization so that we can track them using some identifier... maybe memory address. 
Advantages:
1. Simple to implement
2. Current layer framework can be used
3. current optimizers and their update policy can be used just by adding traits

Disadvantages:
1. although might involve a lot of lookup operations while in training

Solution - (brave and visionary :P ): Each layer will be passed a generator of the update policy at initialization. Each layer will handle its updates separately. Each layer will instantiate the update policy for each parameter and update them accordingly. The common code can be reduced through static polymorphism. A layer has to register its parameters to its static parent and the order of registering the parameters has to be as same as the order in which they are updated thus no lookup is required.
Advantages:
1. No lookup involved, no extra overhead in training
2. Current optimizers and their update policies can be used
Disadvantages:
1. A more robust layer framework has to be formed
2. Extra care has to taken while implementing layers, as described in the description

Last resort: will be discussed when one of the admins closes the issue :P

Any other solutions with different personality traits or changes to the current ones are welcome. Finding a good solution to this problem will significantly increase the performance of the neural network framework. 

Disclaimer: there is a price for suggesting a very brave (too many changes in the current framework), if your suggestion gets accepted, you will pay the price of writing the extra (compared to my brave solution) changes the solution has cost. Just like in Vickrey auction...... haha... don;t worry just kidding. or Am I?

**Comment by zoq** [comment]
> Although elegant, this procedure can be computationally expensive as many matrix transformations are required in transforming each layer's individual matrices to global parameter matrix and vice versa.

Not sure I get what you mean with many matrix transformation, what we basically do is:

```
// global
arma::mat parameterMatrix = arma::mat(100, 100);

// Layer 1
arma::mat layerMatrix = arma::mat(parameterMatrix.memptr(), 10, 10);

// Layer 2
arma::mat layerMatrix = arma::mat(parameterMatrix.memptr() + offset, 10, 10);
```
I'm not sure switching to

```
// Layer 1
arma::mat layerMatrix = arma::mat(10, 10);

// Layer 2
arma::mat layerMatrix = arma::mat(10, 10);
```

would result in any reasonable speedup, maybe I missed something?

The potential speedup I see would be if the complete model doesn't fit in memory. Just wanted to make sure I get all the points right.

**Comment by sumedhghaisas** [comment]
Ahh sorry. I mean in the computation of the gradient. Like in the case of the Linear layer we compute the gradient by 

`
gradient.submat(0, 0, weight.n_elem - 1, 0) = arma::vectorise(
      error * input.t());
`

I am not exactly sure how Armadillo will internally handle this operation. But when I checked the documentation of vectorise it said that it copies the elements. Although if the matrix is temporary, now I think it should not involve any copy, I am not sure. But it will involve a copy operation to the submat view. 

In the other implementation, I was thinking of doing like
`
updatePolicy.Update(weights, error * input.t())
`

Sorry I was not clear in my description before. I thought I was saving 2 copies before, cause vectorize will be used in layers with parameters, but now with the temporary matrix thing, maybe I am just saving 1 copy.

**Comment by sumedhghaisas** [comment]
But yeah, I did not think before about the complete parameter matrix not fitting in the memory. That can also be a factor.

**Comment by scrcro** [comment]
In some cases more optimization can be achieved if only submat of the weights are updated. We need not update all the elements always. But which submat needs to be updated depends on each input, therefore deciding the parameters to be updated upfront at the time of init takes away this flexibility. Word embedding is a case that comes to mind.

**Comment by rcurtin** [comment]
I'm not sure I follow your reasoning here---the current implementation simply allocates the space for all the weights.  It is not required that every layer update all of its respective parameters, so even in the word embedding case you suggest I think it is still possible and ok to avoid updating all elements (depending on the layer types chosen).  The flexibility that we *do* lose as a result of doing this is the ability to resize the parameters matrix during an optimization; but I am not sure that is functionality anyone particularly needs (and it would still be possible to do something strange like that, it would just be backhanded and strange).

**Comment by scrcro** [comment]
For an "embedding layer" the weight matrix could be large, depending on the vocabulary size. While training though, I might train it with a sequence of let's say 10 words. That would mean I only need to update 10 rows in the weight matrix (assuming I have one row for each word). But these 10 rows are not fixed but depends on the input.
As far as I figured from reading the code and playing with it for a while, that all the "parameters" of all the layers are collected during initializing and they are updated depending on the "UpdatePolicy". This way I believe we will be updating all rows of the weight matrix for all layers always. Let me know if I am getting this wrong.

**Comment by rcurtin** [comment]
Okay, so I believe what you are saying is this (correct me if I am wrong):

The optimizer works in such a way that it calls `Gradient()` on the entire network, which returns a dense `arma::mat` containing the gradient values for every parameter in the network (which is stored as a single flattened `arma::mat`).  This means that if a single layer has a sparse gradient (like the embedding layer you suggested), the sparse gradient will be embedded as part of the dense `arma::mat`, and when the actual update is applied by the optimizer, it will not take advantage of the sparsity of the gradient in an individual layer.

That is what I believe you are saying, and if that is what you are saying, that is correct and that is the way the code currently functions.

Given that, I have two thoughts in response:

 * Typically only the first layer will be some layer like an embedding layer with a sparse gradient.  Yes, it is possible that other layers will also have sparse gradients, but I don't think it's wrong to say that in the majority of applications, the gradient will be mostly (very mostly) dense.  Moreover,

 * Sparse matrices can give you less memory footprint and in certain situations, addition and subtraction can be quicker.  But generally the extra overhead required to maintain that sparse matrix means that you only see speedup for extremely sparse matrices in certain cases, and my intuition is that for this case, even if you did go refactor all the code to support sparse gradients for certain layers, it wouldn't actually give you any real speedup.

Let me know if I've misunderstood.

**Comment by scrcro** [comment]
True, embedding layer will be the first layer. The delta and gradient will be a sparse matrix as you say. The embedding matrix though is not going to be sparse. It will be dense but the updates will happen to a few rows only out of the thousands. 
I do not know the internals of armadillo. It may handle this efficiently, if it does then we need not worry. But if it tries to multiply a lot of rows with zero, and repeatedly, for every epoch of the training then this may be a performance issue.
Thoughts..

**Comment by rcurtin** [comment]
This all depends on how the embedding layer (which currently does not exist) is implemented, not necessarily the internals of Armadillo.  As far as I can tell, the abstractions we are using do not rule out the optimizations you are suggesting.

**Comment by rcurtin** [comment]
I'm going to close this issue---to me it's not clear what the actual issue to be solved here is.  If I misunderstood and we should leave it open, maybe we can open a new issue (or multiple issues) which is more about a single specific problem.
